mypy/checkexpr.py

```diff
@@ -613,6 +607,15 @@ def check_call(self, callee: Type, args: List[Expression],
                 on which the method is being called
         """
         arg_messages = arg_messages or self.msg
+
+        if (callable_name is not None and object_type is not None
+                and isinstance(callee, FunctionLike)):
+            # Apply plugin signature hook that may generate a better signature.
+            signature_hook = self.plugin.get_method_signature_hook(callable_name)
+            if signature_hook:
+                callee = self.apply_method_signature_hook(
+                    callee, args, arg_kinds, context, arg_names, object_type, signature_hook)
```

Comment:
I think I understand what you mean - `check_call` is supposed to receive a concrete callable type and check that the given argument list can be applied to that callable type. The code that calls `check_call` has already determined the callable type, so `check_call` shouldn't try to modify it using plugin hooks.

The reason why I put the plugin hook call into `check_call` is because of how method calls are checked in mypy. The usual pattern seems to be:

1. Look up the type of the called method (using `checkexpr.ExpressionChecker.accept` for "real" calls, or `checkmember.analyze_member_access` for implicit calls)
2. Check the actual arguments of the method call against the method type (using `checkexpr.ExpressionChecker.check_call`)

It's not possible to call method signature hooks in the first step - the functions/methods used there are generic to all attribute lookups, not just method calls, so they don't receive the method call-specific information needed by method signature hooks (specifically, information about the actual argument types/values of the call). Since only the second step (`check_call`) has all the necessary information, I put the method signature hook call there.

Maybe it would be better to instead refactor the "look up and apply method signature hook" code into a helper method, and call that manually between steps 1 and 2 where appropriate.

Reply:
> Maybe it would be better to instead refactor the "look up and apply method signature hook" code into a helper method, and call that manually between steps 1 and 2 where appropriate.

This sounds like a promising approach.